### DISCUSSION TITLE
Event Dataflow and keeping data up to date

### DISCUSSION THREAD

**Discussion by Shoggomo** [ASE-Projekte-WS-2021/ase-ws-21-servus#23] [Ideas]
We have multiple places, where we need our Event data kept up to date. 
We need to show new Events on the map and remove non-existing Events. 
We need to update Event details, if a user attends an Event or other data is changed by the host.

To simplify all of this, we could subscribe to updates of the database and inform all activities about updates.
The first part is already implemented, we can subscribe to database changes.
The second part could be implemented by calling an Intent every time data changes. This Intent can be consumed by every activity.

**Implementation notes**:
 - A Service could be used to listen to database updates and send Intents to the rest of the app
   - Service Thread and Intent example: https://stackoverflow.com/a/30629349/2542448
 - The Service should pause to pull data, whenever the app is not active
   - Stop Service, when activity is Paused or stopped example: https://stackoverflow.com/a/26438959/2542448
 - The service needs to be started inside the MainActivity
   - As we rely on BottomSheets and the MainActivity always exists, we need to do this only in this Activity
 - Background Services are deprecated now. The WorkManager seems to be the right way now.
   - https://developer.android.com/reference/androidx/work/WorkManager.html
 - To initially get data, a way to force a fetch of data could be provided
   - This could work via an Intent, that can be sent by any Activity and trigger a fetch

This way, Activities/Fragments only have to listen to a specific Intent and act according to its information.

**Comment by Shoggomo** [comment]
This strategy can also be applied to the data of the Event, that is currently attended, or not attended.
We could have an Intent, that informs the whole application about the state of the "current" Event.
This data can be subscribed to from the database, too. This way we will have fewer problems with stale or wrong data.

**Implementation notes**:
 - When attending an Event:
   - Call backend to increment attendance number (for the MVP; later we upload user information)
   - Save Event ID to SharedPreferences, to remember the Event, if the app is restarted
   - Subscribe to the Event data from the database (this is done in previously mentioned Service)
 - When hosting an Event:
   - Send information to database to create the Event
   - Save Event ID to SharedPreferences, to remember the Event, if the app is restarted
   - Save some information in the SharedPreferences, that we are the host of this Event (we need backend security as well, but thats post MVP)
   - Subscribe to the Event data from the database (this is done in previously mentioned Service)

**Comment by Tiedearnemann** [comment]
I like this :D
My only "concern" might be... Do we really rely ona classic service or rather need to omplement a threat for our work? I ask this as I know, services - especially long-time services - are tended to be killed by the OS pretty fast which can then cause other issues (sync, user experience, performance, etc.)

In the Doc, they state "A service is simply a component that can run in the background, even when the user is not interacting with your application, so you should create a service only if that is what you need. If you must perform work outside of your main thread, but only while the user is interacting with your application, you should instead create a new thread in the context of another application component." (https://developer.android.com/guide/components/services)

**Reply by Shoggomo** [reply]
It seems that Services aren't the proper way to do background tasks anymore, anyway. The current approach seems to use the WorkManager. It looks, like this is the tool for most background tasks now.
